synapse/federation/federation_client.py

```diff
@@ -524,7 +525,17 @@ async def _try_destination_list(
                     e.code,
                     e.args[0],
                 )
+            except ProxiedRequestError as e:
+                # If an auth error comes back from a federation endpoint, pass
+                # it back to the client.
+                if e.code == 403:
+                    raise
+
+                logger.warning(
+                    "Failed to %s via %s", description, destination, exc_info=True
+                )
```

Comment:
how does this work? AFAICT the ProxiedRequestError is raised at line 518, but that won't be caught here.

Reply:
Without this change, the traceback looks like:

<details>

```
2021-04-20 18:22:30,115 - synapse.federation.federation_client - 536 - WARNING - POST-1 - Failed to send_join via hs1
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 754, in _do_send_join
    return await self.transport_layer.send_join_v2(
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/transport/client.py", line 256, in send_join_v2
    response = await self.client.put_json(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 714, in put_json
    response = await self._send_request_with_optional_trailing_slash(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 296, in _send_request_with_optional_trailing_slash
    response = await self._send_request(request, **send_request_args)
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 542, in _send_request
    raise exc
synapse.api.errors.HttpResponseException: 403: Forbidden

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 496, in _try_destination_list
    res = await callback(destination)
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 663, in send_request
    content = await self._do_send_join(destination, pdu)
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 770, in _do_send_join
    raise e.to_synapse_error()
synapse.api.errors.ProxiedRequestError: 403: You do not belong to any of the required spaces to join this room.
```
</details>

This is not catching the exception raised on line 518, but one raised on line 770, i.e. `_do_send_join` has already converting the `HttpResponseException` into a `ProxiedRequestError`. I'm now realizing that this seems specific to the `_do_send_join` use-case, and not all callers of `_try_destination_list`, which is unfortunate. I'm not really sure if `_do_send_join` is doing the right thing or not though. 😢

Reply:
I took a look at this (and put up a change at [`clokep/federation-client-failover`](https://github.com/matrix-org/synapse/compare/develop...clokep/federation-client-failover), but I don't think that's right. It is a bit complicated because there's a few cases here:

* Some of the functions try to support both v1 and v2 endpoints, but only treat some v2 errors as "real" errors, while others are assumed to be non-existent v2 endpoints (and fallback to v1). (The "real" errors get raised as `ProxiedRequestError`.)
* There's a couple of spots where we seem to just raise `SynapseError` directly.
* There seems to be `InvalidResponseError` which should be used in some places instead.
* The only times that an `HttpResponseException` is a 500 error (or if the endpoint doesn't exist) -- this seems to match what I'd expect (the 403 would get passed back to the client properly), but this isn't happening due to it actually be the `ProxiedRequestError` getting raised.

I guess what I'm saying is that this is becoming complicated and I'd like to fix it separately than holding up this change for it. I'm having trouble coming up with the expected behavior for some of these functions though. It might just be to ensure we always return the original `HttpResponseException`, but I suspect that will cause situations where we previously failed over to no longer failover.

Reply:
I think the long and the short of it is that we need to decide whether the conversion from `HttpResponseException` happens at the `do_send_join` level or the `_try_destination_list` level. At the moment, both levels are trying to do it, which leaves us with a mess.

> There's a couple of spots where we seem to just raise SynapseError directly.
> There seems to be InvalidResponseError which should be used in some places instead.

I think the general logic here is that `_try_destination_list` should retry for `InvalidResponseError`, but not for `SynapseError`. But it's very unclear.

> The only times that an HttpResponseException is a 500 error

this sentence seems to be missing some words: I'm unclear what you're saying here.

> I guess what I'm saying is that this is becoming complicated and I'd like to fix it separately than holding up this change for it.

That's fair. I think if you want to take that approach, I'd remove this extra code from `_try_destination_list` for now. The whole "joins don't return clear errors" thing is an existing problem, and we can consider it orthogonal.

Reply:
Thanks! For now I've backed out the changes to `FederationClient` and I'm going to work on them in a separate branch.

Reply:
This seems fairly well documented in the docstring of `_try_destination_list`, TBH I think the callbacks are just doing the wrong thing. But still, will fix it separately.

Reply:
I think #9868 is the fix for this.